### PR TITLE
spatial/r3: add Box methods

### PR DESCRIPTION
**Pull request by soypat**
Please take a look.

Resolves part of https://github.com/gonum/gonum/issues/1791

### CODE REVIEW

File: spatial/r3/vector.go
Comment (sbinet) [inline spatial/r3/vector.go:123]:
```suggestion
// mulElem performs Hadamard product between a vector a
```

File: spatial/r3/vector.go
Comment (sbinet) [inline spatial/r3/vector.go:104]:
```suggestion
// absElem returns a vector with components set to their absolute value.
```

File: spatial/r3/vector.go
Comment (sbinet) [inline spatial/r3/vector.go:113]: I might go with:
```suggestion
// mulElem returns the Hadamard product between the vectors a and b.
```

File: spatial/r3/box.go
Comment (sbinet) [inline spatial/r3/box.go:29]: do we want to document or specify the order?
or document it isn't specified?

Comment (soypat) [reply spatial/r3/box.go:29]: Will document that it is not specified.

**Comment by soypat** [comment]
@kortschak @sbinet Could this and https://github.com/gonum/gonum/pull/1792 be merged sometime this week? I'd like to be using these types for my CERN hackathon application :grinning:

**Review by sbinet — DISMISSED** [review]

**Comment by sbinet** [comment]
although, it'd be great if you could rebase off #1801 (to get the CI running)

**Comment by soypat** [comment]
@kortschak It's all good- thanks for letting me know in advance!

File: spatial/r3/box.go
Comment (kortschak) [inline spatial/r3/box.go:51]: Never put in comments what can go in code.

```suggestion
	return []Vec{
		0: a.Min,
		1: {X: a.Max.X, Y: a.Min.Y, Z: a.Min.Z},
		2: {X: a.Max.X, Y: a.Max.Y, Z: a.Min.Z},
		3: {X: a.Min.X, Y: a.Max.Y, Z: a.Min.Z},
		4: {X: a.Min.X, Y: a.Min.Y, Z: a.Max.Z},
		5: {X: a.Max.X, Y: a.Min.Y, Z: a.Max.Z},
		6: a.Max,
		7: {X: a.Min.X, Y: a.Max.Y, Z: a.Max.Z},
	}
}
```

File: spatial/r3/box.go
Comment (kortschak) [inline spatial/r3/box.go:16]: How often is this needed? It seems to me that it wouldn't be often and that it's simple enough to leave to users. (Though see comment on constructor).

File: spatial/r3/box.go
Comment (kortschak) [inline spatial/r3/box.go:39]: I think this could be clearer.

File: spatial/r3/box.go
Comment (kortschak) [inline spatial/r3/box.go:8]: Can we have a constructor like [`image.Rect`](https://pkg.go.dev/image#Rect)?

File: spatial/r3/box.go
Comment (kortschak) [inline spatial/r3/box.go:78]:
```suggestion
	return a.Min.X <= v.X && v.X <= a.Max.X &&
		a.Min.Y <= v.Y && v.Y <= a.Max.Y &&
		a.Min.Z <= v.Z && v.Z <= a.Max.Z
```

**Review by kortschak — CHANGES_REQUESTED** [review]

**Comment by kortschak** [comment]
Please also rebase onto master.

File: spatial/r3/box.go
Comment (soypat) [reply spatial/r3/box.go:16]: I'd be ok with removing this. That said this form seems to be the most common form of constructing a box from scratch from what I've been doing.
What you are suggesting below is having a `NewBox(a,b Vec)` that ensures valid box creation, right?

Comment (kortschak) [reply spatial/r3/box.go]: > What you are suggesting below is having a `NewBox(a,b Vec)` that ensures valid box creation, right?

Yes, so that `b.Min` is actually `elemMin(a, b)` and `b.Max` is `elemMax(a, b)`.

Comment (soypat) [reply spatial/r3/box.go]: Alright, I've added Empty method along with valid box creation algorithm for `CenteredBox`, `Union` and `Scale`.

I still feel like `CenteredBox` is the most reasonable way of initializing a Box where complexity is reasonably justfied.

I'll also comment on `NewBox`: A `NewBox` constructor which does the same thing as `Rect` is not intuitive to me at all. "NewBox creates a box that contains all components of the two vectors"? In every case where I've created a box I've had precise information on how to create `Min` and `Max` from scratch. I feel like going the `Rect` way lends itself to messy Box creation where the user hopes the resulting Box contains the desired bounds.

Comment (kortschak) [reply spatial/r3/box.go]: The constructor defines the box by the position of its side rather than the position of two of its corners. This is the intuition. An alternative way to think about it is as the union of two zero-sized boxes at the defined min and max (which many not have been min and max).

Comment (soypat) [reply spatial/r3/box.go]: I do not see the difference between defining a box by its side over its corner. The corners lie on the sides after all? 

Let me restate my fear: If we allow constructing valid boxes with two arbitrary vectors then a lot of readability can be lost when it is misused. When I say misuse I mean users can find clever ways of constructing boxes such that in the example`NewBox(a,b)` `a.X < b.X` while `a.Y>b.Y`, resulting in confusion over which vector defines which side/corner (at least for me). When constructing boxes with `Box{Min:a, Max:b}` the intention is clear and the code more structured.

Comment (kortschak) [reply spatial/r3/box.go]: > I do not see the difference between defining a box by its side over its corner. The corners lie on the sides after all?

Taking r2 to reduce complsxit of example, Box{Min:Vec{X:1,Y:1}, Max:Vec{2,2}} does not work the same as Box{Min:Vec{X:1,Y:2}, Max:Vec{2,1}} (this would be an invalid box under the implementations that I was looking at yesterday). In pseudo code Box for X=1..2 Y=1..2 is guaranteed to mean what the user intends.

> If we allow constructing valid boxes with two arbitrary vectors

They are not vectors. Take a look at [`image.Rect`](https://pkg.go.dev/image#Rect); it accepts integer arguments. `NewBox` would be `NewBox(x0, y0, x1, y1 float64) Box`. These parameters merely say make a box with vertical sides at x0 and x1 and horizontal sides at y0 and y1.

Comment (soypat) [reply spatial/r3/box.go]: OK, if we are talking about a function signature like `NewBox(x0, y0, z0, x1, y1, z1 float64)` then I can agree this is OK since now it is clear the box is being created with sides rather than points (as implied by using `Vec` type). I was initially against this signature because it had lots of arguments- however, looking through my use of `Box` it's clear this function would replace **many** `Box{Min...Max...}` initializations.
Little question on the signature: which do we prefer?
```go
NewBox(x0, y0, z0, x1, y1, z1 float64)
NewBox(x0, x1, y0, y1, z0, z1 float64)
```
I am partial to the second one since related numbers are side-by-side.

Comment (kortschak) [reply spatial/r3/box.go]: While I prefer the second, I'm going to argue for the first on the basis of similarity with `image.Rect`.

**Review by kortschak — COMMENTED** [review]
Comments for r2 box apply here.

File: spatial/r3/box.go
Comment (kortschak) [inline spatial/r3/box.go:39]: If this were `Scale(0.5, Add(a.Min, a.Max))` it would always return the centre, even for non-well formed boxes, and it would be infinitesimally cheaper.

File: spatial/r3/box.go
Comment (kortschak) [inline spatial/r3/box.go:9]: One line.

File: spatial/r3/box_test.go
Comment (kortschak) [inline spatial/r3/box_test.go:65]: s/Fatalf/Errorf/

**Review by kortschak — COMMENTED** [review]
Minor changes then LGTM

File: spatial/r3/box.go
Comment (soypat) [reply spatial/r3/box.go]: Good catch

File: spatial/r3/box_test.go
Comment (kortschak) [inline spatial/r3/box_test.go:116]: The addition of a space to make a `Fatal` call format correctly is a smell. I'd prefer `Fatalf` here.

**Review by kortschak — COMMENTED** [review]
LGTM after nit (BTW it's hard to review incrementally with amended changes since the changes are not associated with commits). It's fine to squash as a final move after in principle approval to avoid GitHub's borken approach to squash merge ownership).

**Comment by soypat** [comment]
@kortschak Changes applied without amending since that was probably the last of it. I'll keep that in mind for future commits on other branches.

**Review by kortschak — APPROVED** [review]

**Pull request merged**
